core/node/libp2p.go

```diff
@@ -0,0 +1,591 @@
+}
+
+type BaseRouting routing.IpfsRouting
+type P2PHostOut struct {
+	fx.Out
+
+	Host    host.Host
+	Routing BaseRouting
+	IpfsDHT *dht.IpfsDHT
+}
+
+// TODO: move some of this into params struct
+func P2PHost(mctx MetricsCtx, lc fx.Lifecycle, params P2PHostIn) (out P2PHostOut, err error) {
+	opts := []libp2p.Option{libp2p.NoListenAddrs}
+	for _, o := range params.Opts {
+		opts = append(opts, o...)
+	}
+
+	ctx, cancel := context.WithCancel(mctx)
+	lc.Append(fx.Hook{
+		OnStop: func(_ context.Context) error {
+			cancel()
+			return nil
+		},
+	})
+
+	opts = append(opts, libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+		r, err := params.RoutingOption(ctx, h, params.Repo.Datastore(), params.Validator)
+		out.Routing = r
+		return r, err
+	}))
+
+	out.Host, err = params.HostOption(ctx, params.ID, params.Peerstore, opts...)
+	if err != nil {
+		return P2PHostOut{}, err
+	}
+
+	// this code is necessary just for tests: mock network constructions
+	// ignore the libp2p constructor options that actually construct the routing!
+	if out.Routing == nil {
+		r, err := params.RoutingOption(ctx, out.Host, params.Repo.Datastore(), params.Validator)
+		if err != nil {
+			return P2PHostOut{}, err
+		}
+		out.Routing = r
+		out.Host = routedhost.Wrap(out.Host, out.Routing)
+	}
+
+	lc.Append(fx.Hook{
+		OnStop: func(ctx context.Context) error {
+			return out.Host.Close()
+		},
+	})
+
+	// TODO: break this up into more DI units
+	// TODO: I'm not a fan of type assertions like this but the
+	// `RoutingOption` system doesn't currently provide access to the
+	// IpfsNode.
+	//
+	// Ideally, we'd do something like:
+	//
+	// 1. Add some fancy method to introspect into tiered routers to extract
+	//    things like the pubsub router or the DHT (complicated, messy,
+	//    probably not worth it).
+	// 2. Pass the IpfsNode into the RoutingOption (would also remove the
+	//    PSRouter case below.
+	// 3. Introduce some kind of service manager? (my personal favorite but
+	//    that requires a fair amount of work).
+	if dht, ok := out.Routing.(*dht.IpfsDHT); ok {
+		out.IpfsDHT = dht
+
+		lc.Append(fx.Hook{
+			OnStop: func(ctx context.Context) error {
+				return out.IpfsDHT.Close()
+			},
+		})
+	}
+
+	return out, err
+}
+
+type p2pRoutingIn struct {
+	fx.In
+
+	Repo      repo.Repo
+	Validator record.Validator
+	Host      host.Host
+	PubSub    *pubsub.PubSub `optional:"true"`
+
+	BaseRouting BaseRouting
+}
+
+type p2pRoutingOut struct {
+	fx.Out
+
+	IpfsRouting routing.IpfsRouting
+	PSRouter    *namesys.PubsubValueStore
+}
+
+func P2POnlineRouting(ipnsps bool) func(mctx MetricsCtx, lc fx.Lifecycle, in p2pRoutingIn) (out p2pRoutingOut) {
+	return func(mctx MetricsCtx, lc fx.Lifecycle, in p2pRoutingIn) (out p2pRoutingOut) {
+		out.IpfsRouting = in.BaseRouting
+
+		if ipnsps {
+			out.PSRouter = namesys.NewPubsubValueStore(
+				lifecycleCtx(mctx, lc),
+				in.Host,
+				in.BaseRouting,
+				in.PubSub,
+				in.Validator,
+			)
```

Comment:
So yeah, we could put this in it's own unit then put the code below in it's own unit. That unit would optionally take every kind of router we can construct and, based on the ones we have, construct the composed router.